File: U2/enlace_dinamico/UsoDePilas.cpp
#include "Pila.h"
#include <iostream>

void despedida(Pila *ultimo_elemento) {
  if (ultimo_elemento) {
    cout << "¡Pila no vacia! Se ha vaciado automáticamente" << endl;
    ultimo_elemento = clear(ultimo_elemento);
  }
  cout << "¡Hasta pronto!" << endl;
}

void mostrar_menu() {
  cout << "\nIngrese opcion: " << endl;
  cout << "1. Ingresar elemento a la pila" << endl;
  cout << "2. Mostrar la pila" << endl;
  cout << "3. Quitar el utilmo elemento" << endl;
  cout << "4. Limpiar pila" << endl;
  cout << "5. Salir" << endl;
}

int main() {
  Pila *ultimo_elemento = nullptr;
  int opcion;
  do {
    mostrar_menu();
    cin >> opcion;
    if(!cin) {
      cout << "¡No ingresó dígito válido! ";  
      cin.clear();
      cin.ignore();
    }
    if (opcion == 1)
      ultimo_elemento = push(ultimo_elemento);
    else if (opcion == 2 && ultimo_elemento)
      show(ultimo_elemento);
    else if (opcion == 3 && ultimo_elemento)
      ultimo_elemento = pop(ultimo_elemento);
    else if (opcion == 4 && ultimo_elemento)
      ultimo_elemento = clear(ultimo_elemento);
    else if (opcion == 5)
      despedida(ultimo_elemento);
    else
      cout << "No se realizó ninguna operación" << endl;
  } while (opcion != 5);
  return 0;
}

File: U2/java/app/src/main/java/UsoDePilas.java
import java.util.Scanner;

public class UsoDePilas {

  static LibreriaPila lib = LibreriaPila.INSTANCE;

  public static void main() {
    Scanner escaner = new Scanner(System.in);
    int opcion;
    do {
      mostrarMenu();
      if (!escaner.hasNextInt()) {
        System.out.print("¡No ingresó dígito válido! ");
        opcion = 0;
      } else {
        opcion = escaner.nextInt();
        escaner.nextLine();
      }
      if (opcion == 1) {
        System.out.println("Ingrese el nombre del elemento: ");
        lib.push(escaner.nextLine());
      } else if (opcion == 2 && lib.empty() == 0)
        lib.show();
      else if (opcion == 3 && lib.empty() == 0)
        lib.pop();
      else if (opcion == 4 && lib.empty() == 0)
        lib.clear();
      else if (opcion == 5)
        despedida(lib.empty() == 0);
      else
        System.out.println("No se realizó ninguna operación");
    } while (opcion != 5);
    escaner.close();
  }

  static void despedida(boolean hayElementos) {
    if (hayElementos) {
      System.out.println("¡Pila no vacia! Se ha vaciado automáticamente");
      lib.clear();
    }
    System.out.println("¡Hasta pronto!");
  }

  static void mostrarMenu() {
    System.out.println("\nIngrese opcion: ");
    System.out.println("1. Ingresar elemento a la pila");
    System.out.println("2. Mostrar la pila");
    System.out.println("3. Quitar el utilmo elemento");
    System.out.println("4. Limpiar pila");
    System.out.println("5. Salir");
  }
}
